proof/hash_to_g2.saw

```diff
@@ -0,0 +1,452 @@
+  };
+
+recip_sqrt_fp2_ov <- admit "recip_sqrt_fp2" recip_sqrt_fp2_spec;
+
+let sqrt_align_fp2_unint_spec = do {
+  (ret, ret_ptr) <- ptr_to_fresh "ret" vec384x_type;
+  (sqrt, sqrt_ptr) <- ptr_to_fresh_readonly "sqrt" vec384x_type;
+  (inp, inp_ptr) <- ptr_to_fresh_readonly "inp" vec384x_type;
+  crucible_precond {{ fp2_invariant inp }};
+  crucible_precond {{ fp2_invariant sqrt }};
+  crucible_precond {{ fp2_invariant ret }};
+  crucible_execute_func [ret_ptr, ret_ptr, sqrt_ptr, inp_ptr];
+  crucible_points_to ret_ptr (crucible_term
+      {{ fp2_rep (sqrt_align (fp2_abs ret) (fp2_abs sqrt) (fp2_abs inp)) }});
+  is_sqrt <- llvm_fresh_var "is_sqrt" limb_type;
+  crucible_return (crucible_term {{ is_sqrt }}); // NOTE: return value not used here
+  };
+
+sqrt_align_fp2_unint_ov <-admit "sqrt_align_fp2" sqrt_align_fp2_unint_spec;
+
+// TODO: move somewhere else
+let simp_then_admit r = admit_cryptol (rewrite (cryptol_ss ()) r);
+
+// The optimized version of clear_cofactor given in the IETF spec is correct:
+clear_cofactor_bls12381_correct <- admit_cryptol {{ \p -> Aux::clear_cofactor_bls12381_g2 p == HE2::clear_cofactor p }};
+
+// The SWU map maps to the isogenous curve:
+swu_on_curve <- simp_then_admit {{ \u -> is_point_affine HE2::Isogenous_Curve (HE2::map_to_curve_simple_swu u) == True }};
+
+// point_dadd implements add:
+dadd_thm <- simp_then_admit {{ \ p1 p2 -> affinify HE2::Isogenous_Curve (point_dadd Fp_2 p1 p2 HE2::A' False) == add HE2::Isogenous_Curve (affinify HE2::Isogenous_Curve p1) (affinify HE2::Isogenous_Curve p2) }};
+
+// the addition of two points on the curve is on the curve:
+point_dadd_on_curve <- simp_then_admit {{ \p q -> is_point_projective HE2::Isogenous_Curve (point_dadd Fp_2 p q HE2::A' False) == if is_point_projective HE2::Isogenous_Curve p /\ is_point_projective HE2::Isogenous_Curve q then True else apply is_point_projective HE2::Isogenous_Curve (point_dadd Fp_2 p q HE2::A' False) }};
+
+// If p is a projective point on the isogenous and `affinify p` is point_O, then p's Z coordinate is 0:
+on_curve_is_point_O_lemma <- simp_then_admit {{ \p -> is_point_O HE2::Isogenous_Curve (affinify HE2::Isogenous_Curve p) == if (is_point_projective HE2::Isogenous_Curve p /\ ~is_zero Fp_2 p.2) then False else apply is_point_O HE2::Isogenous_Curve (affinify HE2::Isogenous_Curve p) }};
+
+// The implementation of map_to_curve never returns something of the form (0,0,_) or (_,_,0)
+swu_impl_xy_nz <- simp_then_admit {{ \u -> (~((x,y) == (Fp_2.field_zero,Fp_2.field_zero) \/ z == Fp_2.field_zero) where (x,y,z) = map_to_isogenous_E2_impl u) == True }};
+
+// a relation between is_point_projective and is_point_affine:
+is_point_projective_lemma <- simp_then_admit {{ \p -> is_point_projective HE2::Isogenous_Curve p == if is_point_affine HE2::Isogenous_Curve (affinify HE2::Isogenous_Curve p) /\ ~((p.0, p.1) == (Fp_2.field_zero,Fp_2.field_zero) \/ p.2 == Fp_2.field_zero) then True else apply is_point_projective HE2::Isogenous_Curve p }};
```

Comment:
This deserves a bit more commentary on what it is saying, and why it is true.

Reply:
Added a comment